jsapp/js/components/apiTokenDisplay.tsx

```diff
@@ -0,0 +1,64 @@
+import React, {
+  useState,
+  useEffect,
+} from 'react';
+import bem from 'js/bem';
+import {dataInterface} from 'js/dataInterface';
+import TextBox from 'js/components/common/textBox';
+import Button from 'js/components/common/button';
+
+const HIDDEN_TOKEN_VALUE = '*'.repeat(40);
+
+/**
+ * Displays secret API token of a logged in user.
+ * The token is obfuscated until a "show me" button is clicked.
+ */
+export default function ApiTokenDisplay() {
+  const [token, setToken] = useState(null);
+  const [isFetching, setIsFetching] = useState(false);
+  const [isVisible, setIsVisible] = useState(false);
+
+  const toggleTokenVisibility = () => {
+    setIsVisible(!isVisible);
+  };
+
+  useEffect(() => {
+    if (isVisible && token === null) {
+      const fetchToken = async () => {
+        setIsFetching(true);
```

Comment:
Over-engineering alert :exclamation: probably don't make this change

I believe a more direct translation of setState to hooks would be to use one useState with an object. [Example](https://stackoverflow.com/a/53575023). This allows you to set multiple state objects in command. Just like setState({foo: 1, bar: 2}) would change both foo and bar in one command.

When the user clicks toggle, I would prefer to see one synchronous action result in one state transition. As the state gets more complex, useReducer could be considered (but not in this simple component). 

Action > Synchronous state change > Async Side effect > Synchronous state change.

But all of this philosophy is to deal with more complex state transitions. It helps the developer understand that one action causes X state changes all in one place (rather than searching multiple places for it).

Why I don't think this change is truly necessary to make, is that it will alter your effect usage. Running that effect on every state change, IMO, will be more like class based components. But what you're doing here isn't really wrong. It's justifiable to create multiple states and trigger the effect function only with one state change. Sure you could refactor and throw in a branching statement to only do things on the state changes you care about - but meh.

Reply:
I will not do this then :)